crates/stages/stages/src/stages/era.rs

```diff
@@ -0,0 +1,603 @@
+    N: NodePrimitives<BlockHeader: Value>,
+{
+    fn id(&self) -> StageId {
+        StageId::Era
+    }
+
+    fn poll_execute_ready(
+        &mut self,
+        cx: &mut Context<'_>,
+        input: ExecInput,
+    ) -> Poll<Result<(), StageError>> {
+        if input.target_reached() || self.item.is_some() {
+            return Poll::Ready(Ok(()));
+        }
+
+        if self.stream.is_none() {
+            if let Some(source) = self.source.clone() {
+                self.stream.replace(source.create(input)?);
+            }
+        }
+        if let Some(stream) = &mut self.stream {
+            if let Some(next) = ready!(stream.poll_next_unpin(cx))
+                .transpose()
+                .map_err(|e| StageError::Fatal(e.into()))?
+            {
+                self.item.replace(next);
+            }
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn execute(&mut self, provider: &Provider, input: ExecInput) -> Result<ExecOutput, StageError> {
+        let era = self.item.take().ok_or(StageError::MissingDownloadBuffer)?;
+        let static_file_provider = provider.static_file_provider();
+
+        // Consistency check of expected headers in static files vs DB is done on provider::sync_gap
+        // when poll_execute_ready is polled.
+        let last_header_number = static_file_provider
+            .get_highest_static_file_block(StaticFileSegment::Headers)
+            .unwrap_or_default();
+
+        // Find the latest total difficulty
+        let mut td = static_file_provider
+            .header_td_by_number(last_header_number)?
+            .ok_or(ProviderError::TotalDifficultyNotFound(last_header_number))?;
+
+        // Although headers were downloaded in reverse order, the collector iterates it in ascending
+        // order
+        let mut writer = static_file_provider.latest_writer(StaticFileSegment::Headers)?;
+
+        let height = era::process_iter(
+            era,
+            &mut writer,
+            provider,
+            &mut self.hash_collector,
+            &mut td,
+            last_header_number..=input.target(),
+        )
+        .map_err(|e| StageError::Fatal(e.into()))?;
+
+        self.last_block_height.replace(height);
+
+        if !self.hash_collector.is_empty() {
+            era::build_index(provider, &mut self.hash_collector)
+                .map_err(|e| StageError::Recoverable(e.into()))?;
+            self.hash_collector.clear();
+        }
+
+        Ok(ExecOutput::done(StageCheckpoint::new(height)))
+    }
+
+    fn unwind(
+        &mut self,
+        provider: &Provider,
+        input: UnwindInput,
+    ) -> Result<UnwindOutput, StageError> {
+        self.hash_collector.clear();
+        self.item.take();
+        self.last_block_height.take();
+        self.stream.take();
+
+        if let Some(source) = &self.source {
+            // Bodies
+            stages::ensure_consistency(provider, Some(input.unwind_to))?;
```

Comment:
do we even need to do anything on unwind here, isn't all of this already covered by the header+body stages anyway?

Reply:
The stages should know how to revert themselves. By doing this we're taking away that ability from `EraStage` and making it dependent on other stages, even though the API does not strictly suggests that's the case.

But since `EraStage` is optional and only supplements the work of other stages, I suppose we can make an exception.